File: app/src/main/java/com/nimtego/plectrum/domain/interactor/MainViewInteractor.java
package com.nimtego.plectrum.domain.interactor;

import com.nimtego.plectrum.presentation.main.model.AlbumModel;
import com.nimtego.plectrum.presentation.main.model.AlbumModelK;
import com.nimtego.plectrum.presentation.main.model.ArtistModel;
import com.nimtego.plectrum.presentation.main.model.ArtistModelK;
import com.nimtego.plectrum.presentation.main.model.MainDataModel;
import com.nimtego.plectrum.presentation.main.model.MainDataModelK;
import com.nimtego.plectrum.presentation.main.model.SongModel;
import com.nimtego.plectrum.presentation.main.model.SongModelK;

import java.util.List;

import dagger.internal.Preconditions;
import io.reactivex.Observable;

public class MainViewInteractor
        extends BaseInteractor<MainDataModelK, MainViewInteractor.Params> {

    public MainViewInteractor() {
    }

    @Override
    protected Observable<MainDataModelK> buildUseCaseObservable(MainViewInteractor.Params params) {
        Preconditions.checkNotNull(params);
        Observable<List<AlbumModelK>> albumsObs = repository.albums(params.request);
        Observable<List<SongModelK>> songsObs = repository.songs(params.request);
        Observable<List<ArtistModelK>> artistsObs = repository.artists(params.request);
        return Observable.combineLatest(albumsObs, songsObs, artistsObs, (albums, songs, artists) ->
                new MainDataModelK(artists, albums, songs));
    }

    public static final class Params {

        private final String request;

        private Params(String request) {
            this.request = request;
        }

        public static Params forRequest(String request) {
            return new Params(request);
        }
    }

}

File: app/src/main/java/com/nimtego/plectrum/domain/interactor/MainViewInteractor.kt
package com.nimtego.plectrum.domain.interactor

import com.nimtego.plectrum.presentation.main.model.MainDataModel
import dagger.internal.Preconditions
import io.reactivex.Observable
import io.reactivex.functions.Function3


class MainViewInteractor : BaseInteractor<MainDataModel, MainViewInteractor.Params>() {

    override fun buildUseCaseObservable(params: MainViewInteractor.Params): Observable<MainDataModel> {
        Preconditions.checkNotNull(params)
        val albumsObs = repository.albums(params.request)
        val songsObs = repository.songs(params.request)
        val artistsObs = repository.artists(params.request)
        //todo to ugly
        return Observable.combineLatest(
                    albumsObs,
                    songsObs,
                    artistsObs,
                    Function3 { albums,
                                songs,
                                artists ->
                        MainDataModel(artists, albums, songs)
                })
    }

    class Params private constructor(val request: String) {
        companion object {

            fun forRequest(request: String): Params {
                return Params(request)
            }
        }
    }

}